### ISSUE TITLE
bug in inference script?

### ISSUE THREAD

**Issue by lingcong-k** [DeepSceneSeg/EfficientPS#17]
HI, thanks for the great repo.

I am doing simple testing with ur script cityscapes_inference.py
it worked well with cityspace images while the size is 1024*2048
but for othersize images,it is somehow not correct.

it has a doubled results with the instance part and the semantic head results not overlapped with each other..
so ended up with two results separate.

the problems lies in there where the mask_pred prediction positions are wrong (not to ref size)
```
            bbx_inv = invert_roi_bbx(det_bboxes[:, :4], 
                      tuple(mask_pred.shape[2:]), ref_size)
            bbx_idx = torch.arange(0, det_bboxes.size(0), 
                      dtype=torch.long, device=det_bboxes.device)
            
            mask_pred = roi_sampling(mask_pred, bbx_inv, bbx_idx, 
                        ref_size, padding="zero")
```

![image](https://user-images.githubusercontent.com/73646719/114586694-3c3ac500-9c85-11eb-8f0d-d556a4494120.png)

am a bit confused by the script, looks like an easy bug, but havent found a proper way to solve it. could you please take a look? Thanks

**Comment by mohan1914** [comment]
Can you try again? I added a line to interpolate semantic logits to ref _size. It should ensure both semantic and instance logits are of the same shape. 
If it doesn't work, can you add the image shape details that you work with so that I can replicate the issue.

**Comment by lingcong-k** [comment]
HI, thanks for the updates. I tried with ur newest updates.. it is still the same issue.

The image i used is this one: https://specials-images.forbesimg.com/imageserve/1220958526/960x0.jpg?fit=scale
Probably u can test with this..

Also I notice there's difference between the results i run with ur latest checkpoint and the one run on ur online demo:
for example this image from cityscapes
![image](https://user-images.githubusercontent.com/73646719/114736296-4375d780-9d46-11eb-8d19-3f9783fee069.png)

from your online demo, it has results like this:
![image](https://user-images.githubusercontent.com/73646719/114736400-59839800-9d46-11eb-9752-e691f9db1650.png)

But if I run with cityspace_inference.py with the config and model downloaded from your cityscapes checkpoint (http://panoptic.cs.uni-freiburg.de/static/models/efficientPS_kitti.zip). I got results like this:
![image](https://user-images.githubusercontent.com/73646719/114737127-ffcf9d80-9d46-11eb-8aaa-707cdbf28512.png)

Those small person seems not as good as ur demo results. is there any postprocessing steps am missing here?
Thanks

**Comment by mohan1914** [comment]
Thanks for pointing out the issue. I will try to debug it when I have some free time. A quick and dirty workaround will be to resize the image such that the height is 1024 while maintaining the aspect ratio. Get the panoptic output and resize it back to the original shape.
[Panoptic Image Output](https://drive.google.com/file/d/1MVT-qgSdBJ-DAuJfpnSDqzpYst_E5DEZ/view?usp=sharing)

For the demo predictions, we use Mapillary pre-trained model with multi-scale + flip test time augmentations.

**Comment by lingcong-k** [comment]
> Thanks for pointing out the issue. I will try to debug it when I have some free time. A quick and dirty workaround will be to resize the image such that the height is 1024 while maintaining the aspect ratio. Get the panoptic output and resize it back to the original shape.
> [Panoptic Image Output](https://drive.google.com/file/d/1MVT-qgSdBJ-DAuJfpnSDqzpYst_E5DEZ/view?usp=sharing)
> 
> For the demo predictions, we use Mapillary pre-trained model with multi-scale + flip test time augmentations.

yeah thanks. i did the hack

**Comment by lingcong-k** [comment]
> Thanks for pointing out the issue. I will try to debug it when I have some free time. A quick and dirty workaround will be to resize the image such that the height is 1024 while maintaining the aspect ratio. Get the panoptic output and resize it back to the original shape.
> [Panoptic Image Output](https://drive.google.com/file/d/1MVT-qgSdBJ-DAuJfpnSDqzpYst_E5DEZ/view?usp=sharing)
> 
> For the demo predictions, we use Mapillary pre-trained model with multi-scale + flip test time augmentations.

Hi, I have a question regarding the checkpoints.

The pretrained weight we download for cityscape is only trained on cityscapes dataset? training set 2975, val 500

but for the best performance one on cityscapes, u trained with mapillary vistas and cityscapes together?  18k training set from mapillary + cityscapes 2975

May I also know how long does it take to train the best performance model (i read that u trained with 16 TitanX)?

Thanks :)

**Comment by 48313758a** [comment]
When I run cityscapes_inference.py, I cannot get the result I want. Is cityscapes_inference.py needed to add show_result()?
this is my result with result = inference_detector(model, os.path.join(path, imgName), eval='panoptic'):
[[tensor([[3, 3, 3,  ..., 2, 2, 2],
        [3, 3, 3,  ..., 2, 2, 2],
        [3, 3, 3,  ..., 2, 2, 2],
        ...,
        [1, 1, 1,  ..., 1, 1, 1],
        [1, 1, 1,  ..., 1, 1, 1],
        [1, 1, 1,  ..., 1, 1, 1]]), tensor([255,   0,   4,   8]), [{'filename': 'imgs/test/2_leftImg8bit.jpg', 'ori_shape': (1024, 1396, 3), 'img_shape': (384, 524, 3), 'pad_shape': (384, 544, 3), 'scale_factor': 0.375, 'flip': False, 'img_norm_cfg': {'mean': array([106.433, 116.617, 119.559], dtype=float32), 'std': array([65.496, 67.6  , 74.123], dtype=float32), 'to_rgb': True}}]]]

**Comment by mohan1914** [comment]
We have updated the readme to reflect the purpose of the two saving scripts:
- tool/cityscapes_inference.py: saves predictions in the official cityscapes panoptic format.
- tool/cityscapes_save_predictions.py: saves color visualizations. 

if you are looking for visualization you should use the second script.

**Comment by mohan1914** [comment]
@lingcong-k Due to multi-machine multi-gpu training the full mapillary+cityscapes training usually took about a week.
